src/main/java/codeu/model/store/basic/UserStore.java

```diff
@@ -121,4 +128,101 @@ public boolean isUserRegistered(String username) {
   public void setUsers(List<User> users) {
     this.users = users;
   }
+
+  /**
+    * Returns the number of users currently recorded
+    */
+  public int getUserCount() {
+    return users.size();
+  }
+
+  /**
+    * Returns the users array
+    */
+  public List<User> getAllUsers(){
+    return users;
+  }
+
+  /**
+    * Returns the user with the most messages sent
+    * Pre : the the users list should already have the messageCount calculated
+    */
+  public String getTopUser(){
+    sortUserList();
+    return users.get(0).getName();
+  }
+
+  /**
+    * Iterate through all messages and get the count for the owner of ther messages
+    * Should only be called in PersistentDataStore when the messageCount attr does not exist
+    */
+  public int getMessageCountForUser(UUID uuid){
+    List<Message> messages = MessageStore.getInstance().getAllMessages();
+    int messageCount = 0;
+    for (Message message : messages) {
+      UUID id = message.getAuthorId();
+      if (id != null) {
+        if (uuid == id) {
+          messageCount++;
+        }
+      }
+    }
+    return messageCount;
+  }
+
+  /**
+    * Sort the users list by using sort in the library
+    */
+  public void sortUserList(){
+    Collections.sort(users, new Comparator<User>(){
+      public int compare(User a, User b){
+        return b.getMessageCount() - a.getMessageCount();
+      }
+    });
+  }
+
+  /**
+    * make a JSON file that consists JSON objects {user, messageCount}
+    * sample JSONFile :
+    * [
+    *   {messageCount : 20, name : userOne},
+    *   {messageCount : 10, name : userTwo}
+    * ]
+    */
+  public void writeJSON(){
+    // get the users array sorted with the corresponding messageCount setup
+    sortUserList();
+
+    try {
+      // check if the directory exists, if not, create it
+      File path = new File("./api/stats");
```

Comment:
I think this will have concurrency issues if multiple users are trying to run the function at the same time.
You can consider having writeJson() take an argument for the file name.  That way each load of the page can generate a UUID and have it's own file which is written.

Another option is to expose this json as the response to an API.  The javascript can make an ajax call for the json.

Reply:
so something like writeJSON(UUID filename){}; ? and ill make the UUID my filename instead of the userData.json

Will this not end up being new files created every time a user reloads but the old files are never deleted?

Reply:
Yes.  But not an UUID id you already have such as userId, since in theory the same user can load the page multiple times concurrently.  It needs to be a freshly generated one.

I think you can also leverage java's temporary file libraries.  It looks like there's also a deleteOnExit() to handle the case you're talking about.  https://docs.oracle.com/javase/7/docs/api/java/io/File.html#deleteOnExit%28%29

I haven't looked into the library closely, but most of the time the library provides an interface to create a temporary file.  set that temporary file to delete.  retrieve the path of the generated temporary file.  you can use the retrieved path for writing the file contents as well as access the file in your jsp.